File: practice-problems/is-balanced.js
const openers = ["(", "[", "{"]
const closers = [")", "]", "}"]
let braketOrder = []

const isBalanced = (str) => {
  if (str.length % 2 !== 0) {
    return false
  }

  for (let i = 0; i < str.length; i++) {
    // charToMatch = str.charAt(i)
    if (closers.includes(str.charAt(i)) && closers.indexOf(str.charAt(i)) === openers.indexOf(braketOrder[braketOrder.length - 1]) ) {
      braketOrder.pop()
    } else if (openers.includes(str.charAt(i))) {
      braketOrder.push(str.charAt(i))
    } else {
      return false
    }
  }
  return true
}

File: practice-problems/is-balanced.rb
def isBalanced(string)

  # iterate forward through string
  # track opening order
  # check closing order
      # impossible to fail on an opener until you reach the end of the string
      # can only fail on a closer
      # so simply track the openers and only scrutinize the closers

  openers = ["(", "[", "{"]

  unpairedOpeners = []

  string.split('').each do |char|
      
      if openers.include?(char)
          # opener -- track
          unpairedOpeners << char
      else
          # closer -- evaluate
          partner = unpairedOpeners.pop
          case char
              when ")"
                  if partner != "("
                      return false
                  end
              when "]"
                  if partner != "["
                      return false
                  end
              when "}"
                  if partner != "{"
                      return false
                  end
              else
                  return nil
          end
      end
      
  end

  if unpairedOpeners.size == 0
      return true
  else
      return false
  end

end

puts isBalanced("()[]{}")